### DISCUSSION TITLE
Using all Kyber KEM implmementation versions (Reference, AVX2, 90s)

### DISCUSSION THREAD

**Discussion by yamus-didine** [strongswan/strongswan#1818] [Q&A] [answered]
Hello.

In my project, I am experimenting Kyber KEM with IPsec. I am using strongswan 6.0.0beta3 with liboqs 0.7.1 

I configured liboqs as follows
`cmake -GNinja -DOQS_USE_OPENSSL=ON -DBUILD_SHARED_LIBS=ON -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=Release -DOQS_BUILD_ONLY_LIB=ON`

I configured strongswan as follows
`./configure --prefix=/usr --sysconfdir=/etc --disable-ikev1 --enable-frodo --enable-oqs --enable-silent-rules`

I managed to use Kyber KEM in the initial key exchange by editing the file `src/libstrongswan/crypto/proposal/proposal.c` in strongswan.

My inquiry is that only AVX2 implementation is enabled when using the mentioned setup.

I want to be able to use all kyber implementations (reference, 90s, avx2). Here is what I did

1. I edited `*oqs_kem_create ( )` in  `src/libstrongswan/plugins/oqs/oqs_kem.c`

![oqs_kem_create](https://github.com/strongswan/strongswan/assets/134303261/89e7d657-f391-49b0-b912-7c2acd990366)

2. I edited `METHOD (...)` in `src/libstrongswan/plugins/oqs/oqs_plugin.c` 

![oqs_plugin](https://github.com/strongswan/strongswan/assets/134303261/1941736e-91ce-4105-a23f-2a3c00fb417b)

3. I edited `key_exchange_is_kem(...)` in  `src/libstrongswan/crypto/key_exchange.c`

![key_exchange _is_kem](https://github.com/strongswan/strongswan/assets/134303261/768a48ce-b058-4b8c-af62-e648bd076d76)

4.  I edited `ENUM_NEXT(...)` in  `src/libstrongswan/crypto/key_exchange.c`

![Enum_next](https://github.com/strongswan/strongswan/assets/134303261/2058c097-c265-48db-8997-8d8dbc55a263)

5.  I added the short keyword `kyber19` which stands for **Kyber level 1, 90s version** in `ENUM_NEXT(...)` in  `src/libstrongswan/crypto/key_exchange.c`

![keyword_short](https://github.com/strongswan/strongswan/assets/134303261/983196d6-baf6-4e56-a291-c1ed31121c78)

6. I edited `enum_key_exchange_method_t` in  `src/libstrongswan/crypto/key_exchange.h`

![enum_key_exchange_method](https://github.com/strongswan/strongswan/assets/134303261/4066f15c-8e6c-47a1-af35-797f8ad8d217)

7.  I edited `proposal_add_supported_ike(...)` in  `src/libstrongswan/crypto/proposal/proposal.c`

![add_enumerator](https://github.com/strongswan/strongswan/assets/134303261/c3f46b5e-2b2b-479b-b809-3bb3180973b7)

8. I added the line 203 in the file `src/libstrongswan/crypto/proposal/proposal_keywords_static.txt`

![keywords_static_txt](https://github.com/strongswan/strongswan/assets/134303261/e87c5d41-8ee7-471f-bf4e-1160c7555dc9)

9. I edited `proposal_toke_wordlist[ ]` in the file `src/libstrongswan/crypto/proposal/proposal_keywords_static.c`

![keywords_static_c](https://github.com/strongswan/strongswan/assets/134303261/76989426-e2c9-4d2d-b004-c446efde5d4c)

After I made the edits mentionned above, I made `make clean && make uninstall` in strongswan folder. i then recompiled strongswan

I set the IKE proposals in one of the hosts as `proposals = aes128gcm128-sha256-kyber19`. I used the keyword `kyber19`. However, I got this log when I loaded `charon`

![charon](https://github.com/strongswan/strongswan/assets/134303261/46287520-8f99-4b88-8a77-35ad42998299)

I would also need to integrate the reference implementation as well. I did not know how to do it similarly to the attempted work I did with the 90s version.

I need help please !!

**Comment by tobiasbrunner** [comment, marked as answer]
> I managed to use Kyber KEM in the initial key exchange by editing the file `src/libstrongswan/crypto/proposal/proposal.c` in strongswan.

Why? There should be no need to edit that to use it as initial KE if you configure an appropriate proposal.

> I want to be able to use all kyber implementations (reference, 90s, avx2).

For what purpose?

Also note that the 90s variants seem to be alternatives and, if enabled in liboqs (see `KYBER_90S` define), replace the default variants (i.e. a separate identifier doesn't make much sense as you have to recompile liboqs to use one or the other). They also have been removed with liboqs 0.8.0 (according to the [release notes](https://github.com/open-quantum-safe/liboqs/releases/tag/0.8.0)).


Don't know exactly what's going wrong with your modifications, but other than in the _oqs_ plugin you don't need to change any of these files to add a new key exchange method. Just register the new keyword(s) via `lib->proposal->register_token()`. Then you can configure your proposal with it.
